project-api/ballerina-projects/src/main/java/io/ballerina/projects/writers/BaloWriter.java

```diff
@@ -29,18 +59,347 @@
  */
 public class BaloWriter {
 
+    private static final String MODULES = "modules";
+
+    private BaloWriter() {}
+
     /**
      * Write a package to a .balo and return the created .balo path.
-     * @param pkg Package to be written as a .balo.
+     *
+     * @param pkg  Package to be written as a .balo.
      * @param path Directory where the .balo should be created.
      * @return Newly created balo path
      */
     public static Path write(Package pkg, Path path) {
-        Path balo = path.resolve("org-name-any-version.balo");
         // todo check if the given package is compiled properly
-        // todo check if the path is a directory
+
+        // Check if the path is a directory
+        if (!path.toFile().isDirectory()) {
+            throw new RuntimeException("Given path is not a directory: " + path);
+        }
+
         // Check directory permissions
-        // Handle any io errors
+//        if (!new File(String.valueOf(path)).canWrite()) {
+//            throw new NoPermissionException("No write access to create balo:" + path);
+//        }
+
+        if (!path.toFile().canWrite()) {
+            throw new NoPermissionException("No write access to create balo:" + path);
+        }
+
+        BallerinaToml ballerinaToml = pkg.ballerinaToml();
+        Path packagePath = pkg.packagePath();
+        Path balo = path.resolve(getBaloName(ballerinaToml));
+
+        // Create the archive over write if exists
+        try (FileSystem baloFS = createBaloArchive(balo)) {
+            // Now lets put stuff in
+            populateBaloArchive(baloFS, packagePath, ballerinaToml);
+        } catch (IOException e) {
+            throw new BLangCompilerException("Failed to create balo :" + e.getMessage(), e);
+        } catch (BLangCompilerException be) {
```

Comment:
We can write the `Files.delete(balo);` within the first catch block (before L99). This way we don't need to catch and re-throw the BLangCompilerException

Reply:
`IOException` in L 98. associate to the error when creating balo. At that case, we cant do `Files.delete(balo);` since balo is not created.